Checkout: Hide contact validation errors during auto-completion, with external state

sirbrillig opened the pull request:
#### Proposed Changes

When checkout loads, it fetches saved contact details from an API endpoint, pre-fills the contact form with those details, and attempts to auto-complete the contact details step. Since https://github.com/Automattic/wp-calypso/pull/64344, this process uses the same logic as if a user filled out the form and clicked "Continue" manually. 

However, there is a downside of this approach: the contact details are validated every time someone clicks "Continue" and since this uses the same process, if the saved details are invalid the user will see an error reported as soon as checkout loads. Since the user didn't actually take any action at this point, this can be very confusing.

<img width="977" alt="validation-error" src="https://user-images.githubusercontent.com/2036909/201235660-6c218f93-0ae4-418d-bb69-29e953fe404f.png">

In this PR, we refactor the placement of the auto-completion code so that its validation call will not report any errors on failure. This is done by moving the auto-completion into the parent component of the `CheckoutStepGroup` where the validation function is run and refusing to show a validation error message if auto-completion is still in-progress.

Fixes https://github.com/Automattic/wp-calypso/issues/69211

#### Differences compared to the previous version of this PR

This is the second attempt at solving this issue after https://github.com/Automattic/wp-calypso/pull/69943 was reverted in https://github.com/Automattic/wp-calypso/pull/70721 because there was a bug with free purchases. 

In the previous PR, we moved the checkout contact details step into a wrapper component due to a leaky abstraction. There is one case where no contact details step is needed: when the purchase is free ($0 total, not full credits). Before https://github.com/Automattic/wp-calypso/pull/69943, we did not render the step in this case, but after that PR, we _did_ render the step, except that the wrapper returned null. The result was that, if the purchase was free, checkout would think that the contact details step was present and _active_ even though it was null, effectively blocking checkout.

This PR performs additional refactoring to remove the leaky abstraction and gets rid of the wrapper. In addition, https://github.com/Automattic/wp-calypso/pull/70776 adds a test to prevent the bug from happening again.

Some of the changes originally made for this PR have been merged separately with https://github.com/Automattic/wp-calypso/pull/70776 (which improves the tests) and https://github.com/Automattic/wp-calypso/pull/70778 (which refactors checkout steps to externalize the store).

#### Testing Instructions

There are unit tests for this behavior but you can test it manually as follows:

First test that successful validation still works:

- Make sure you have valid cached _domain_ contact details for a user (using a domain product will cause more fields to be visible). You can do this by visiting checkout when a domain product is in your cart, filling out the contact information step and pressing "Continue". Assuming they are valid, they will be saved for future visits to checkout.
- Add a _domain_ product to your cart and visit checkout (or reload if you are already there).
- Verify that you see the contact information step load before it is auto-completed.
- Verify that the active step is now the payment method step.

Next test that manual validation displays error messages:

- Add a _domain_ product to your cart and visit checkout.
- Click to edit the contact information step if it is not already active.
- Enter invalid contact information (eg: by clearing the "Last name" field).
- Press "Continue".
- Verify that you see a validation error.
- Verify that the active step is still the contact information step.

Finally, test that auto-completion does not display error messages:

- Apply D92139-code and sandbox the API. This will break the domain contact details validation endpoint.
- Add a _domain_ product to your cart and visit checkout.
- Verify that you see the contact information step load before it is auto-completed.
- Verify that you _do not see_ a validation error.
- Verify that the active step is still the contact information step.

sirbrillig committed:
Add shouldWait and isComplete to useCachedDomainContactDetails

`shouldWait` is an argument that can be used to delay fetching contact
details (and therefore also to delay filling the contact form).

`isComplete` is a return value to let the caller know that the contact
form has been auto-completed (or at least completion has been
attempted).

Because this changes the arguments, this is a breaking change.

sirbrillig committed:
Remove useCachedDomainContactDetails call inside WPContactForm

sirbrillig committed:
Update signature of useCachedDomainContactDetails in tests

sirbrillig committed:
Move contact step into custom step component

We want to call `useCachedDomainContactDetails` above `WPContactForm` so
that we can determine when the contact form auto-fill validation is
in-progress and silence the validation errors during that time.

However, to have access to step completion the hook must be called
inside `CheckoutStepGroup`, and we will need its return value for the
`isCompleteCallback` prop of the contact form step.

Therefore, we must put a wrapper component inside `WPCheckout` to call
the hook and use its return value for the contact form step.

sirbrillig committed:
Do not show contact validation errors if auto-completion is running

sirbrillig committed:
Add test to prove that autocompletion does not show errors

sirbrillig committed:
Do not render CheckoutContactStep at all if contactDetailsType none

sirbrillig committed:
Change useCachedContactDetails to return null if shouldWait true

sirbrillig committed:
Add logstash when useCachedContactDetailsForCheckoutForm has an error

sirbrillig committed:
Fix types of contactDetailsType in CheckoutContactStep

sirbrillig committed:
Revert creation of custom checkout step component

sirbrillig committed:
Revert unncessary shouldWait mod to useCachedDomainContactDetails

sirbrillig committed:
Use CheckoutStepGroupStore directly in useCachedDomainContactDetails

sirbrillig committed:
Refactor useCachedDomainContactDetails to use step store

sirbrillig committed:
Update CheckoutStepGroup/useCachedDomainContactDetails in tests

sirbrillig committed:
Move some calls in WPCheckout back to their original position

sirbrillig committed:
Prevent unmounted side effects in useCachedDomainContactDetails

sirbrillig committed:
Remove typo in checkout-contact-step test added by rebase

spraveenitpro commented:
<details>

<summary>Running unit tests</summary>



###  client/my-sites/checkout/composite-checkout/test/checkout-contact-step.js

Tests passed with a warning:

![eeCUvv.png](https://user-images.githubusercontent.com/552016/205738237-61b92d64-4896-40d0-b4ae-a1eb68d60c94.png)

### client/my-sites/checkout/composite-checkout/test/checkout-contact-step.js

**Tests passed with a warning:**

![qOB0xB.png](https://user-images.githubusercontent.com/552016/205738589-d964b2aa-f741-419a-8653-0d04905d0cdd.png)

</details>

<details>

<summary>Running functional tests</summary>

<details>

<summary>First test that successful validation still works and verify that the active step is now the payment method step:</summary>

**Works as described.**

![vz6q8l.png](https://user-images.githubusercontent.com/552016/205740493-b8a3ae20-59fe-47f2-87c4-9f412dbad8ba.png)
</details>
<details>
<summary>Manual validation displays error messages. Verify that the active step is still the contact information step</summary>

**Works as described.**

![5q8uVv.png](https://user-images.githubusercontent.com/552016/205741045-3e96b52c-a99a-435b-842e-23430eeaa4db.png)
</details>
<details>
<summary>Auto-completion does not display error messages. Verify that the active step is still the contact information step.:</summary>

**Works as described.**

![S83rkU.png](https://user-images.githubusercontent.com/552016/205741564-029e7073-4899-4407-8e44-18a69604b491.png)
</details>
</details>

pottedmeat commented on client/my-sites/checkout/composite-checkout/components/wp-checkout.tsx:
```diff
@@ -237,6 +239,13 @@ export default function WPCheckout( {
 	const [ is3PDAccountConsentAccepted, setIs3PDAccountConsentAccepted ] = useState( false );
 	const [ isSubmitted, setIsSubmitted ] = useState( false );
 
+	const stepGroupStore = useMemo( () => createCheckoutStepGroupStore(), [] );
```
The empty `deps` threw me off a little. (No change expected).

pottedmeat reviewed (approved):
Changes makes sense, code looks good, type checking I did all looks good.

sirbrillig commented:
> Running into a problem with the unit test. CheckoutStepGroupInner is calling setTotalSteps which calls onStateChange which runs the callback in store.subscription.subscribe which calls setContextValue which triggers a redraw in the current execution loop.

Good point.

> If I make the call to onStateChange async (e.g. setTimeout( onStateChange, 0 );) the test passes, though I don't know what side-effects that might create.

Interesting idea. I tried it in two places, `onStateChange` and inside `notifySubscribers` (which seems maybe more appropriate?) but in both cases the unit tests fail (`yarn test-packages composite-checkout`). This may just be an issue where the unit tests are expecting the UI to update immediately and instead it takes an additional render, but my hope was that this change would modify as little behavior as possible.

I wonder if it's possible to just not call `onStateChange` inside `setTotalSteps`? Originally, `setTotalSteps` happened in an async call itself, but it wasn't fast enough, so I made it run on every render and short-circuit if there are no changes. Maybe there's no longer any need to re-render if the total number of steps changes since that's more of an internal bookkeeping thing.

sirbrillig committed:
Use user-event library for testing composite-checkout events

sirbrillig committed:
Defer state updates when step group store changes

sirbrillig commented:
Nope, that broke the auto-complete. I have another idea, to defer the call to `setState` itself. That seems to work.

sirbrillig committed:
Update Package lock for user-event

sirbrillig committed:
Also add testing-library/dom to composite-checkout

jjchrisdiehl commented:
Ran through the test steps and after loading the cart with a domain subscription, the 'Enter your contact information' step auo-completes and loads without error and the cart awaits me to press Continue. Upon pressing Continue, the Last Name field fails validation (as expected due to D92139-code).

Unit tests also pass without any warnings.

sirbrillig committed:
Change useCachedContactDetails to only return when fetch complete

This is really three changes in one:

1. Export useCachedContactDetails.
2. Return isLoading as part of its return value.
3. Consider it loading and do not return data until the data has been
   fetched.

sirbrillig committed:
Add shouldWait to useCachedContactDetailsForCheckoutForm

sirbrillig committed:
Remove unnecessary useCachedDomainContactDetails

sirbrillig committed:
Delay loading CheckoutMain until useCachedContactDetails returns

sirbrillig committed:
Call useCachedContactDetailsForCheckoutForm directly in WPCheckout

sirbrillig committed:
Rename cached-contact-details file to be more accurate

sirbrillig committed:
Fix imports of cached-contact-details

sirbrillig committed:
Always mock contact and validation endpoints in contact step tests

sirbrillig committed:
Make shouldWait optional in useCachedContactDetailsForCheckoutForm

sirbrillig committed:
Update cached-contact-details tests for new functions

sirbrillig commented:
I was seeing an unexpected delay when checkout loaded while the contact details were fetched; during the delay, the contact form was editable and active, and had no evidence that a fetch was taking place until it suddenly autocompleted the form. I think this delay actually exists in production too, but may only show up when the contact details endpoint is very slow.

To fix it, I did two things:

1. I refactored `useCachedContactDetails` to always wait until fetching was complete before returning its data.
2. I changed checkout to display a loading page until the cached contact details fetching is complete.

sirbrillig committed:
Restore guard preventing autocomplete if steps are not rendered

sirbrillig commented:
Oh, but now the original bug has returned. This is quite a roller coaster. 😓

sirbrillig commented:
Closing in favor of https://github.com/Automattic/wp-calypso/pull/71024
